crates/primitives/snos-output/src/codec.rs

```diff
@@ -0,0 +1,80 @@
+use alloc::vec::Vec;
+
+use starknet_api::hash::StarkFelt;
+
+use crate::{MessageL1ToL2, MessageL2ToL1, StarknetOsOutput};
+
+/// This codec allows to convert structured OS program output into array of felts
+/// 
+/// In order to prepare parameters for the Starknet contract `updateState` method:
+///     1. Cast the output to dynamic uint256[] array
+///     2. Get onchain data hash & size
+///     3. ABI encode parameters 
+///  
+/// NOTE: Field element (252 bit) is encoded as an EVM word (256 bit) and vice versa
+/// EVM developer should be aware of that and prevent data loss by not using the higest 4 bits
+pub trait SnosCodec: Sized {
+    fn felt_len(&self) -> usize;
+    fn append_to(self, output: &mut Vec<StarkFelt>);
+
+    fn as_vec(self) -> Vec<StarkFelt> {
+        let mut output: Vec<StarkFelt> = Vec::with_capacity(self.felt_len());
+        self.append_to(&mut output);
+        output
+    }
+}
+
+fn segment_append_to<T: SnosCodec>(segment: Vec<T>, output: &mut Vec<StarkFelt>) {
+    let segment_size = segment.iter().map(|msg| msg.felt_len() as u64).sum::<u64>();
+    output.push(segment_size.into());
+    for item in segment {
+        item.append_to(output);
+    }
+}
```

Comment:
You should 
```rust
impl SnosCodec for Vec<T>
  where T: SnosCodec
{
  ...
}
```  

And the reuse it instead of 
```
segment_append_to(self.messages_to_l1, output);
segment_append_to(self.messages_to_l2, output);
```
and

```
output.push((self.payload.len() as u64).into());
let mut payload = self.payload;
output.append(&mut payload);
```

Reply:
It's a bit different for segment and Vec<StarkFelt>, in the fist case we deal with dynamically sized elements